File: web/js/lib/onboarding-steps.js
/**
 * Onboarding step machine (WP 4a.6) — pure decision logic for the 3-step
 * first-run flow ported from docs/design/vault-app-mockup.html's onboarding
 * (frozen round 7), adapted to the DECIDED web-UI shape
 * (docs/WORKPACKAGES.md Phase 4a header; ADR-0004 addendum):
 *
 *   Step 1 Connect  — vault name (optional) + vault API key + a real
 *                      server/key check (no server-URL/connection-profile
 *                      fields: the web app is served BY vault-api and the
 *                      WP 4a.1 CSP's `connect-src 'self'` makes any other
 *                      origin unusable, so there is nothing to pick — see
 *                      api.js's module header for where `getServerUrl`/
 *                      `setServerUrl` used to live).
 *   Step 2 Steam    — optional: Steam Web API key + SteamID64 (the relay,
 *                      WP 4a.6r), never Valve OpenID (that is the native
 *                      Android app's device-local path, ADR-0004 decision 2,
 *                      untouched here).
 *   Step 3 Done     — summary, then close.
 *
 * Kept pure (no DOM, no fetch) so the gating rules are testable head-on:
 * step 1 cannot be left until the key has actually been verified against a
 * live server (`tested: true`) — unlike the mockup, which only gated on
 * form completeness, this fork's step 1 result is used for real (the
 * verified key is what onboarding.js stores), so advancing past it on an
 * unverified guess would silently ship a broken API key into localStorage.
 * Step 2 is optional by design (mockup: "Continue without one — you can
 * sign in later under Settings") and step 3 is terminal.
 */

export const STEP = Object.freeze({ CONNECT: 1, STEAM: 2, DONE: 3 });
export const FIRST_STEP = STEP.CONNECT;
export const LAST_STEP = STEP.DONE;

const STEP_TITLES = Object.freeze({
  [STEP.CONNECT]: "Step 1 of 3 · Connect",
  [STEP.STEAM]: "Step 2 of 3 · Steam identity",
  [STEP.DONE]: "Step 3 of 3 · Ready",
});

/** @param {number} step */
export function stepTitle(step) {
  return STEP_TITLES[step] || "";
}

/** @param {number} step @returns {number} 33 | 66 | 100 */
export function progressPercent(step) {
  const clamped = clampStep(step);
  return Math.round((clamped / LAST_STEP) * 100);
}

/** @param {number} step */
export function clampStep(step) {
  if (!Number.isFinite(step)) return FIRST_STEP;
  return Math.min(LAST_STEP, Math.max(FIRST_STEP, Math.trunc(step)));
}

/**
 * Can the flow leave `step` right now?
 * @param {number} step
 * @param {{tested?: boolean}} state
 */
export function canAdvance(step, state) {
  if (step === STEP.CONNECT) return state?.tested === true;
  return true; // Steam identity (step 2) is optional; step 3 has no "next"
}

/**
 * @param {number} step current step
 * @param {{tested?: boolean}} state
 * @returns {number} the step to move to (unchanged if blocked)
 */
export function nextStep(step, state) {
  if (!canAdvance(step, state)) return step;
  return clampStep(step + 1);
}

/** Mockup rule: step 1 has no Back — `prevStep(1, ...)` is a no-op. */
export function prevStep(step) {
  return clampStep(step - 1);
}

/**
 * Should the onboarding overlay be shown at all? First run only: no vault
 * API key stored yet, and the user has not already chosen demo mode
 * (mockup: "Skip for now — browse in demo mode" is itself an onboarding
 * exit, not a state that reopens it).
 * @param {{hasApiKey: boolean, demoMode: boolean}} ctx
 */
export function shouldShowOnboarding({ hasApiKey, demoMode }) {
  return !hasApiKey && !demoMode;
}

File: app/app/src/main/java/dev/steamvault/app/ui/onboarding/logic/OnboardingSteps.kt
package dev.steamvault.app.ui.onboarding.logic

/**
 * Onboarding step machine (WP 4b.7) -- pure decision logic for the 3-step
 * first-run flow, ported from `web/js/lib/onboarding-steps.js` (WP 4a.6,
 * itself ported from `docs/design/vault-app-mockup.html`'s frozen round-7
 * onboarding), adapted to what THIS app's step 1 actually needs to collect
 * -- a connectivity profile choice, a base URL, and a vault API key (the
 * mockup's original shape), unlike the web port's step 1, which dropped the
 * server-URL/profile fields entirely because the web app is served BY
 * vault-api itself (`web/js/lib/onboarding-steps.js`'s own module header
 * explains why that shortcut does not apply here -- a native app has no
 * "already being served by the thing it is connecting to" starting point).
 *
 *   Step 1 Connect — connectivity profile (System-VPN vs. public domain) +
 *                    base URL + vault API key, gated on a REAL, verified
 *                    connection test (`net/connection/ConnectionCheck.kt`) --
 *                    same as the web port's step 1, [canAdvance] refuses to
 *                    leave this step on an unverified guess.
 *   Step 2 Steam    — optional: Steam OpenID sign-in (this app's OWN
 *                     device-local flow, WP 4b.3 -- unlike the web port,
 *                     which uses the vault-api Steam relay for identity
 *                     too). WP 4h.4 removed this step's OTHER original
 *                     half -- a device-local Steam Web API key entry field
 *                     -- entirely: library data is relayed through
 *                     vault-api unconditionally once signed in, so there is
 *                     no key left for this step to collect.
 *   Step 3 Done     — summary, then the caller persists to
 *                     [dev.steamvault.app.storage.CredentialStore] and
 *                     shows the real app.
 *
 * WP APP-DEMO adds one more exit besides finishing step 3: "Skip for now —
 * browse in demo mode" (steps 1/2 only, mirroring the web port's own
 * demo-skip link) leaves the flow without a vault-api connection at all,
 * using [dev.steamvault.app.demo.DemoState] fixtures instead — see
 * [shouldShowOnboarding]'s own kdoc for how that interacts with this
 * module's original "first run only" rule.
 *
 * Kept pure (no Android framework, no network) so the gating rule is
 * testable head-on, same reasoning as the web source: step 1 cannot be left
 * until [OnboardingController]'s connection test has actually succeeded
 * against a live server, not merely "the fields are non-blank" -- advancing
 * on an unverified guess would silently persist a broken connection into
 * [dev.steamvault.app.storage.CredentialStore].
 */
enum class OnboardingStep { CONNECT, STEAM, DONE }

val FIRST_ONBOARDING_STEP = OnboardingStep.CONNECT
val LAST_ONBOARDING_STEP = OnboardingStep.DONE

private val ORDER = OnboardingStep.entries

/** @return 33 | 66 | 100 -- matches the web port's [progressPercent] rounding. */
fun onboardingProgressPercent(step: OnboardingStep): Int =
    Math.round((step.ordinal + 1) * 100.0 / ORDER.size).toInt()

/**
 * Can the flow leave [step] right now? Mirrors
 * `onboarding-steps.js::canAdvance` exactly: step 1 requires [tested], every
 * other step has nothing blocking it (step 2 is optional by design -- the
 * WP brief's "skippable" boundary; step 3 is terminal, `next` is a no-op
 * there per [nextOnboardingStep]).
 */
fun canAdvanceOnboardingStep(step: OnboardingStep, tested: Boolean): Boolean = when (step) {
    OnboardingStep.CONNECT -> tested
    OnboardingStep.STEAM, OnboardingStep.DONE -> true
}

/** @return the step to move to (unchanged if [canAdvanceOnboardingStep] refuses, or already last). */
fun nextOnboardingStep(step: OnboardingStep, tested: Boolean): OnboardingStep {
    if (!canAdvanceOnboardingStep(step, tested)) return step
    val next = step.ordinal + 1
    return if (next < ORDER.size) ORDER[next] else step
}

/** Mockup rule (mirrored from the web port): step 1 has no Back -- a no-op there. */
fun previousOnboardingStep(step: OnboardingStep): OnboardingStep {
    val prev = step.ordinal - 1
    return if (prev >= 0) ORDER[prev] else step
}

/**
 * Should the onboarding flow be shown at all? Mirrors the web port's own
 * `shouldShowOnboarding({hasApiKey, demoMode})` exactly (WP APP-DEMO):
 * first run only -- no working vault-api connection configured yet AND
 * demo mode has not already been chosen. [demoMode] is itself an
 * onboarding EXIT (`MainActivity`'s "Skip for now -- browse in demo mode"
 * action), not a state that reopens it, same as the web port's own
 * "Skip for now" comment documents.
 */
fun shouldShowOnboarding(hasVaultConnection: Boolean, demoMode: Boolean): Boolean =
    !hasVaultConnection && !demoMode
